hms-api/src/repository/hackathon-repository.ts

```diff
@@ -32,12 +34,39 @@ export async function listHackathons(): Promise<Hackathon[]> {
 }
 
 export async function putHackathon(hackathon: Hackathon) {
+  if (await hackathonExists(hackathon.id)) {
+    throw new InvalidStateError('Cannot create Hackathon, it already exists');
+  }
+
+  if (await hackathonSlugExists(hackathon.slug)) {
+    throw new InvalidStateError('Cannot create Hackathon, slug already exists');
+  }
```

Comment:
these are only an issue if you are trying to create a hackathon, but not when you want to edit one so they shouldn't go here

Reply:
now that I've read more, this comment only applies to the `hackathonExists` one.  But likely you need to check that the hackathon title exists... I'll have to dive deeper into this though]

Reply:
so your saying dynamodb automatically checks that it never randomly choose the same uuid?

Reply:
not exactly... so if we consider the flow for creating a new hackathon (or any entity really):

1. request comes into the rest layer with data
2. data is passed to the service layer where a new hackathon is created
  a. the constructor for that hackathon sets new random UUID as the ID of the hackathon
3. the hackathon is checked for validity
4. the hackathon is passed to the repository layer
  a. at this point we can be relatively sure that the ID does not exist in DynamoDB because it is a random UUID
6. we add the hackathon into DynamoDB

so for the case where we make a new one we don't really need to check because of how UUIDs work, we CAN check but it is a bit overkill tbh.  Here is a nice quote about UUIDs:

> [As per Wikipedia](https://en.wikipedia.org/wiki/Universally_unique_identifier), the number of UUIDs generated to have > atleast 1 collision is 2.71 quintillion.
> This is equivalent to generating around 1 billion UUIDs per second for about 85 years.

If we then consider the flow for updating it is mostly the same, and in that flow we want that ID to be in DynamoDB already so that we can update it's values.  This way we can have a single function that we call in the repository to fulfill both purposes.

It's always good to try and get re-use out of a function when possible (and within reason).  If you have to start adding tons of arguments to the function so that it can do all sorts of things, that is a bit of a code smell.  But in these two scenarios it really just has to do the same thing so we can get re-use out of it